### DISCUSSION TITLE
Add `turbo ls`

### DISCUSSION THREAD

**Discussion by ntgussoni** [vercel/turborepo#379] [Help] [answered]
### Describe the feature you'd like to request

With lerna, we have `lerna list` or `lerna ls`. See: https://github.com/lerna/lerna/tree/main/commands/list

This basically lets you list packages and information in a parseable manner and also provides filtering so that you can, for example, do the following:

`lerna ls --since origin/master` And it will list all packages that changed since a specific commit.

This is useful for pipelines since sometimes you don't want to build and deploy everything unless it actually changed.

I couldn't find an alternative with the available commands today, so I'm just doing it through a bash script but I suppose it's best to streamline it through turborepo just to make sure everything is coherent.

### Describe the solution you'd like

Whatever mimics lerna, for my use-case is fine. The important part is the filtering in my case since otherwise it just lists all available packages.

### Describe alternatives you've considered

N/A

**Comment by ntgussoni** [comment]
#271 might be related

**Comment by ntgussoni** [comment]
I want to add that "changed" is a little bit more complex than the diff between commits, because if A is a dependency of B and A changes, then B should be considered changed even though not a single bit of code is different in that package

**Comment by mishabruml** [comment]
I'd like to see this feature please 🙏

**Comment by jenseng** [comment, marked as answer]
Seems like you can more or less accomplish this today with something along these lines:

```shell
turbo build --dry=json '--filter=...[main]' | jq -r '.packages[]'
```

Which outputs something like:
```
//
some-changed-package
another-changed-package
this-package-didnt-change-directly-but-it-depends-on-some-changed-package-so-there-you-go
...
```

To break it down:
1. `build --dry` does a dry run of the `build` pipeline, and the output will show even those packages without a `build` script (note that it doesn't have to be `build`, any pipeline could do)
2. `=json` formats the dry run output as json
3. for the filter:
   * `...` means include dependencies (e.g. if package B depends on A, but only A changed, include both B and A)
   * `[main]` means check for changes relative to `main` ... swap this out for another ref/sha as appropriate for your use case
4. we then pipe it through `jq` to output just the package names
